crypto/ml_dsa/ml_dsa_sign.c

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/rand.h>
+#include "ml_dsa_local.h"
+#include "ml_dsa_key.h"
+#include "ml_dsa_params.h"
+#include "ml_dsa_matrix.h"
+#include "ml_dsa_sign.h"
+#include "ml_dsa_hash.h"
+
+#define ML_DSA_MAX_LAMBDA 256 /* bit strength for ML-DSA-87 */
+
+/*
+ * @brief Initialize a Signature object by pointing all of its objects to
+ * preallocated blocks. The values passed for hint, z and
+ * c_tilde values are not owned/freed by the |sig| object.
+ *
+ * @param sig The ML_DSA_SIG to initialize.
+ * @param hint A preallocated array of |k| polynomial blocks
+ * @param k The number of |hint| polynomials
+ * @param z A preallocated array of |l| polynomial blocks
+ * @param l The number of |z| polynomials
+ * @param c_tilde A preallocated buffer
+ * @param c_tilde_len The size of |c_tilde|
+ */
+static void signature_init(ML_DSA_SIG *sig,
+                           POLY *hint, uint32_t k, POLY *z, uint32_t l,
+                           uint8_t *c_tilde, size_t c_tilde_len)
+{
+    vector_init(&sig->z, z, l);
+    vector_init(&sig->hint, hint, k);
+    sig->c_tilde = c_tilde;
+    sig->c_tilde_len = c_tilde_len;
+}
+
+/*
+ * FIPS 204, Algorithm 7, ML-DSA.Sign_internal()
+ * @returns 1 on success and 0 on failure.
+ */
+static int ml_dsa_sign_internal(ML_DSA_CTX *ctx, const ML_DSA_KEY *priv,
+                                const uint8_t *encoded_msg,
+                                size_t encoded_msg_len,
+                                const uint8_t *rnd, size_t rnd_len,
+                                uint8_t *out_sig)
+{
+    int ret = 0;
+    const ML_DSA_PARAMS *params = ctx->params;
+    EVP_MD_CTX *h_ctx = ctx->h_ctx;
+    uint32_t k = params->k, l = params->l;
+    uint32_t gamma1 = params->gamma1, gamma2 = params->gamma2;
+    uint8_t *alloc = NULL, *w1_encoded;
+    size_t w1_encoded_len = 128 * k;
```

Comment:
The length `w1_encoded_len` depends on gamma2, and is actually `192 * k` in `ML-DSA-44`.